### PR TITLE
r1.13-rc0 cherry-pick request: fix races in TF-TRT integration

### PR DESCRIPTION
**Pull request by aaroey**
This fixes two data races and one conversion bug in TF-TRT integration:

1. during INT8 calibration, when two sess.run()s fetch two different sets of ops and both of them trigger the computation of the TRTEngineOp, two different TRTEngineOp will be created, and dev_tensors_ will only be allocated once by the first op by the device's resource manager via LookupOrCreate(). As a result, the second TRTEngineOp will have a empty dev_tensors_ vector which will result in std::out_of_range error when the op tries to access it later.
2. IExecutionContext::enqueu() is not thread-safe which cause one of the issues mentioned in
https://github.com/tensorflow/tensorflow/issues/23853. This PR added a mutex to guard that call.
3. IConstantLayer doesn't set the output data type automatically but defaults to FP32, this cause conversion of models that contain any compatible INT32 binary ops to fail. This PR fix that by explicitly setting the output datatype.

### CODE REVIEW

**Comment by aaroey** [comment]
@pooyadavoodi @trevor-m Just FYI, please help to take a look.

File: tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc [hunk 1]
```diff
@@ -402,9 +402,12 @@ bool TRTEngineOp::ExecuteTrtEngine(
                                                 ->implementation()
                                                 ->GpuStreamMemberHack()));
 
+  // nvinfer1::IExecutionContext::enqueue is not thread safe and we need a mutex
```
Comment (pooyadavoodi) [inline tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc]: I would rephrase this to the following:

nvinfer1::IExecutionContext::enqueue can be used by multiple threads simultaneously as long different threads use different execution contexts. But since we use one execution context for all threads, we guard the runtime by a mutex. Since nvinfer1::IExecutionContext::enqueue is an asynchronous function, this guard will be safe only if all the threads use the same CUDA stream which is the current behavior in Tensorflow.

Comment (aaroey) [reply tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc]: Thanks. Since this is a comment change, instead of added a separate commit for that to the PR, I'll fix it in another commit to head.

**Comment by aaroey** [comment]
@aselle Hi Andrew, I just double checked, the CI build failure is irrelevant, please let me know if there are any questions. Thanks.

**Review by aselle — APPROVED** [review]

**Comment by aaroey** [comment]
@aselle I just double checked again, all the failures of required checks are not caused by this PR, just FYI.

**Comment by aaroey** [comment]
Thanks @aselle for the merge!
